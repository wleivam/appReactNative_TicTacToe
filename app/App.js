/**
 * @format
 * @flow strict-local
 */

import React, { useState, useEffect } from 'react';
import type {Node} from 'react';
import {
    Container,
    InfoBar,
    Header,
    Body,
    Content,
    Footer,
} from './components/commons';
import Game from './components/Game';
import Board from './components/Board';
import { players, keyPositions } from './utils';

const App: () => Node = () => {
    const [moves, setMoves] = useState(0);
    const [data, setData] = useState(Array(9).fill(''));
    const [player, setPlayer] = useState(0);
    const [status, setStatus] = useState('playing');

    // Function for make a move in the game
    const play = (position) => {
        setData(newData => {
            newData.splice(position, 1, players[player]);
            return [...newData];
        });
        setMoves(newMoves => {
            newMoves += 1;
            return newMoves;
        });
    };

    // Function for reset the game
    const reset = () => {
        if ( moves > 0) {
            setData(Array(9).fill(''));
            setStatus('playing');
            setPlayer(0);
            setMoves(0);
        }
    };

    // Hook for validate the game state
    useEffect(() => {
        if ( moves > 0 ) {
            let playing = true;
            if ( moves >= 5 ) {
                keyPositions.forEach(positions => {
                    const winner = [...new Set(positions.map(index => data[index]))];
                    if ( winner.length === 1 && winner[0] !== '' ) {
                        setStatus('winner');
                        playing = false;
                        return;
                    }
                });
            }
            if ( moves === 9 && playing ) {
                setStatus('draw');
                playing = false;
            }
            if (playing) setPlayer(player ? 0 : 1);
        }
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [moves]);

    return (
        <Container>
            <InfoBar />
            <Header />
            <Body>
                <Content>
                    <Game
                        player={player}
                        reset={reset}
                        moves={moves}
                        status={status}
                    />
                    <Board
                        data={data}
                        play={play}
                        status={status}
                    />
                </Content>
            </Body>
            <Footer />
        </Container>
    );
};

export default App;
