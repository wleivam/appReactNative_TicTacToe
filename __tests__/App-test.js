/**
 * @format
 */

import 'react-native';
import React from 'react';
import App from '../app/App';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('App initial render', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
});
