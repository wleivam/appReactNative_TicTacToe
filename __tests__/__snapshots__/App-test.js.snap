// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App initial render 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {},
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "height": "10%",
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "height": "80%",
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "height": "90%",
            "width": "80%",
          },
        ]
      }
    >
      <View
        className="sc-bdvvtL ggDfgN"
      >
        <Text
          className="sc-gsDKAQ fyGiCt"
          testID="game-status"
        >
          Next Player: X
        </Text>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="reset-button"
        >
          <Text
            className="sc-hKwDye btRsZv"
          >
            Reset
          </Text>
        </View>
      </View>
      <RCTScrollView
        CellRendererComponent={[Function]}
        ItemSeparatorComponent={[Function]}
        contentContainerStyle={
          Object {
            "height": "100%",
          }
        }
        data={
          Array [
            "",
            "",
            "",
            "",
            "",
            "",
            "",
            "",
            "",
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEnabled={false}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        style={
          Array [
            Object {},
          ]
        }
        viewabilityConfigCallbackPairs={Array []}
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                  },
                  Object {
                    "flex": 1,
                  },
                ]
              }
            >
              <View
                bothBorder={false}
                leftBorder={false}
                rightBorder={true}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "borderRightColor": "#000",
                      "borderRightStyle": "solid",
                      "borderRightWidth": 2,
                      "display": "flex",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <View
                  accessibilityState={
                    Object {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#DDD",
                      "borderRadius": 20,
                      "height": "85%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "85%",
                    }
                  }
                  testID="cell-touchable-00"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 20,
                          "fontWeight": "900",
                        },
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <View
                bothBorder={true}
                leftBorder={false}
                rightBorder={false}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "borderLeftColor": "#000",
                      "borderLeftStyle": "solid",
                      "borderLeftWidth": 2,
                      "borderRightColor": "#000",
                      "borderRightStyle": "solid",
                      "borderRightWidth": 2,
                      "display": "flex",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <View
                  accessibilityState={
                    Object {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#DDD",
                      "borderRadius": 20,
                      "height": "85%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "85%",
                    }
                  }
                  testID="cell-touchable-01"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 20,
                          "fontWeight": "900",
                        },
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <View
                bothBorder={false}
                leftBorder={true}
                rightBorder={false}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "borderLeftColor": "#000",
                      "borderLeftStyle": "solid",
                      "borderLeftWidth": 2,
                      "display": "flex",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <View
                  accessibilityState={
                    Object {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#DDD",
                      "borderRadius": 20,
                      "height": "85%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "85%",
                    }
                  }
                  testID="cell-touchable-02"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 20,
                          "fontWeight": "900",
                        },
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "borderColor": "#000",
                    "borderStyle": "solid",
                    "borderWidth": 2,
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                  },
                  Object {
                    "flex": 1,
                  },
                ]
              }
            >
              <View
                bothBorder={false}
                leftBorder={false}
                rightBorder={true}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "borderRightColor": "#000",
                      "borderRightStyle": "solid",
                      "borderRightWidth": 2,
                      "display": "flex",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <View
                  accessibilityState={
                    Object {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#DDD",
                      "borderRadius": 20,
                      "height": "85%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "85%",
                    }
                  }
                  testID="cell-touchable-03"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 20,
                          "fontWeight": "900",
                        },
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <View
                bothBorder={true}
                leftBorder={false}
                rightBorder={false}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "borderLeftColor": "#000",
                      "borderLeftStyle": "solid",
                      "borderLeftWidth": 2,
                      "borderRightColor": "#000",
                      "borderRightStyle": "solid",
                      "borderRightWidth": 2,
                      "display": "flex",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <View
                  accessibilityState={
                    Object {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#DDD",
                      "borderRadius": 20,
                      "height": "85%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "85%",
                    }
                  }
                  testID="cell-touchable-04"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 20,
                          "fontWeight": "900",
                        },
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <View
                bothBorder={false}
                leftBorder={true}
                rightBorder={false}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "borderLeftColor": "#000",
                      "borderLeftStyle": "solid",
                      "borderLeftWidth": 2,
                      "display": "flex",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <View
                  accessibilityState={
                    Object {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#DDD",
                      "borderRadius": 20,
                      "height": "85%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "85%",
                    }
                  }
                  testID="cell-touchable-05"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 20,
                          "fontWeight": "900",
                        },
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "borderColor": "#000",
                    "borderStyle": "solid",
                    "borderWidth": 2,
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                  },
                  Object {
                    "flex": 1,
                  },
                ]
              }
            >
              <View
                bothBorder={false}
                leftBorder={false}
                rightBorder={true}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "borderRightColor": "#000",
                      "borderRightStyle": "solid",
                      "borderRightWidth": 2,
                      "display": "flex",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <View
                  accessibilityState={
                    Object {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#DDD",
                      "borderRadius": 20,
                      "height": "85%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "85%",
                    }
                  }
                  testID="cell-touchable-06"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 20,
                          "fontWeight": "900",
                        },
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <View
                bothBorder={true}
                leftBorder={false}
                rightBorder={false}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "borderLeftColor": "#000",
                      "borderLeftStyle": "solid",
                      "borderLeftWidth": 2,
                      "borderRightColor": "#000",
                      "borderRightStyle": "solid",
                      "borderRightWidth": 2,
                      "display": "flex",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <View
                  accessibilityState={
                    Object {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#DDD",
                      "borderRadius": 20,
                      "height": "85%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "85%",
                    }
                  }
                  testID="cell-touchable-07"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 20,
                          "fontWeight": "900",
                        },
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <View
                bothBorder={false}
                leftBorder={true}
                rightBorder={false}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "borderLeftColor": "#000",
                      "borderLeftStyle": "solid",
                      "borderLeftWidth": 2,
                      "display": "flex",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <View
                  accessibilityState={
                    Object {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#DDD",
                      "borderRadius": 20,
                      "height": "85%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "85%",
                    }
                  }
                  testID="cell-touchable-08"
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 20,
                          "fontWeight": "900",
                        },
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "height": "10%",
          "width": "100%",
        },
      ]
    }
  />
</RCTSafeAreaView>
`;
